course4/homework-unique-pointer/test.cc

```diff
@@ -0,0 +1,89 @@
+#include<iostream>
+#include<cstring>
+#include <cassert>
+#include "collectable.h"
+#include "refcounted.h"
+
+template<typename T, typename Arg>
+void innerCall(Arg arg1, Arg arg2)
+{
+    auto c1 = course::CollectablePointer<T>::createInstance(arg1);
+    auto c2 = course::CollectablePointer<T>::createInstance(arg2);
+    auto c3 = c1;
+    auto c4 = std::move(c2);
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 0);
+}
+
+template<typename T, typename Arg>
+void test(Arg arg1, Arg arg2)
+{
+    innerCall<T, Arg>(arg1, arg2);
+
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 3);
+    course::CollectablePointer<T>::collect();
+    assert(course::CollectablePointer<T>::itemsToCollectCount() == 0);
+}
+
+template<typename T, typename Arg>
+void refcauntTest(Arg arg)
+{
+     auto c1 = course::RefcountingPointer<T>::createInstance(arg);
+     auto c2 = c1;
+     auto c3 = c2;
+     assert(*c1 == *c3);
+     assert(c1.getCount() == 3);
+     auto c4 = std::move(c2);
+     assert(*c4 == *c3);
+     assert(c1.getCount() == 3);
+     c2 = c1;
+     assert(c1.getCount() == 4);
+     c3 = c1;
+     assert(c1.getCount() == 4);
+     c2 = std::move(c4);
+     assert(c1.getCount() == 3);
```

Comment:
Good test, but incomplete.

Create an class that modifies some global variable (or variables) in constructor and destructor. This way you could verify if object is destroyed when you expect it to be.